### PR TITLE
[Form] Add time zone translations

### PR DESCRIPTION
**Pull request by c960657**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #13662
| License       | MIT
| Doc PR        | 

This is a second attempt to localise `Symfony\Component\Form\Extension\Core\Type\TimezoneType`. First attempt was #17628 that was rejected, because it changed the existing UI a bit, and because I was not fully aware of how to use the CLDR data.

I really think localisation is important. Not everyone speaks English, and if you visit e.g. a Russian-language site, you would expect the timezone picker to show city names written in Cyrillic letters. Let's find a way to fix this, even if it requires adjusting the existing UI slightly.

In this attempt I have tried to stay closer to the existing UI than in #17628. One intentional change, though, is that “Argentina - San Juan” is now displayed as “San Juan, Argentina”. CLDR uses this formatting for US states, e.g. “Tell City, Indiana”, and I think it makes sense, so that list items can be sorted alphabetically by the city name.

One improvent over the existing solution is that `Antarctica/DumontDUrville` is now displayed as “Dumont d’Urville” rather than “DumontDUrville”.

Matching the current UI is a bit tricky, because CLDR does not contain translations for the oceans (Pacific/Atlantic/Indian), so they must be provided manually. But this is necessary, if we want to preserve the existing UI. One way to get around this (and simplify the code a bit) is to group timezones by continent as shown in [this list](https://www.unicode.org/cldr/charts/33/supplemental/territory_containment_un_m_49.html), rather than by the tzdata area. This will give nearly identical results, except that timezones in `Atlantic/xxx`, `Pacific/xxx` and `Indian/xxx` will be shown under the continent they belong to, and a few other exceptions (e.g. Turkey now belongs to Asia, while tzdata puts it in Europe, because Istanbil is there – see https://unicode.org/pipermail/cldr-users/2018-April/000795.html). These are all minor changes IMO.

I have chosen to base the implementation on the excellent Punic library rather than downloading the XML files directly, like I did in #17628, or trying to add them to the Intl component, like I did in #17636. The Punic library supports large parts of CLDR (the data and the specification on how to use it), so I think it is a good choice. It is not a runtime dependency, but is only used for generating the translation files,  so I hope adding this dependency is not too controversial.

_Disclaimer: I was recently [invited](https://github.com/punic/punic/issues/114) to become a co-author of Punic, because I had contributed a bit to the library. However, I do not have any interests in this library, except as an interested user and contributor._

One thing still missing from this PR is that timezones are not sorted according to their localised name. Stack Overflow tells me that in order to do that, one should translate them and sort them prior to adding them to the form. This is e.g. when `CountryType` does. However, I'm not sure about the best way to do this (i.e. how to inject the translator). Another option is to use Punic as a runtime dependency to get the localised strings without using the translator, but I don't know if Symfony wants this kind of dependencies. It could be made as an optional dependency with fallback to the  English names like to today. So I kindly ask for advice on how to handle this.

### CODE REVIEW

**Comment by mvrhov** [comment]
IMO punic should not be required. We have intl and the translations should be taken from there and this should be the "hard" requirement for translating. Otherwise we are just making a mess out of it.

**Comment by c960657** [comment]
> We have intl and the translations should be taken from there and this should be the "hard" requirement for translating.

I previously suggested adding timezone names to the Intl component in #17636. One comment I received was this:

> There are several databases that are kept up to date independently that can provide this data. Why burden Symfony with it?

To be fair, the patch had other problems due to my then limited understanding of CLDR.


CLDR is an excellent source of translation data, and I think it does not get the attention it deserves from the PHP community. I think Symfony should embrace a library that exposes CLDR data, either included by expanding the Intl component or by requiring an external library. However, the Intl component is currently very limited in this respect, and bringing it on par with e.g. Punic would require a substantial amount of work.

But if there is consensus that this data should live in the Intl component, I will reopen #17636.

**Comment by mvrhov** [comment]
with intl I have meant the extension itself. This extension provides the translations from CLDR. Also all this translations are already shipped with symfony!

**Comment by c960657** [comment]
True, the intl PHP extension contains the localised city names. AFAICT it does not contain names of countries and continents, though, but these few strings can be provided manually in translation files.

> Also all the translations from intl (ICU) component are available and are already shipped with Symfony!
I'm not sure what you mean. The localised city names are not included in Symfony?

File: src/Symfony/Component/Form/Resources/translations/generated/timezones.ar.xlf [hunk 1]
```diff
@@ -0,0 +1,1726 @@
+<?xml version="1.0" encoding="utf-8"?>
+<xliff xmlns="urn:oasis:names:tc:xliff:document:1.2" version="1.2">
+  <file source-language="en" target-language="ar" datatype="plaintext" original="file.ext">
+    <header>
+      <tool tool-id="update-timezone-translations.php" tool-name="CLDR time zone translation generator"/>
+    </header>
+    <body>
+      <trans-unit id="xKNxrNf" resname="Africa">
```
Comment (hhamon) [inline src/Symfony/Component/Form/Resources/translations/generated/timezones.ar.xlf]: These ids are autogenerated?

**Comment by mvrhov** [comment]
Sorry for the confusion then. Then my suggestion would be to put this into the intl component as it is requred to localize other form types.

**Review by ro0NL — COMMENTED** [review]
for the implementation, i think we should follow same the approach as other intl types. Thus a  `TimeZoneBundle`

while at it could solve #22302 too

File: src/Symfony/Component/Form/composer.json [hunk 2]
```diff
@@ -33,7 +33,8 @@
         "symfony/http-kernel": "~3.4|~4.0",
         "symfony/security-csrf": "~3.4|~4.0",
         "symfony/translation": "~3.4|~4.0",
-        "symfony/var-dumper": "~3.4|~4.0"
+        "symfony/var-dumper": "~3.4|~4.0",
+        "punic/punic": "~3.1"
```
Comment (ro0NL) [inline src/Symfony/Component/Form/composer.json]: not sure we need this? see e.g. https://3v4l.org/HOv3k

**Comment by ro0NL** [comment]
See #28831 for an alternative

**Comment by ro0NL** [comment]
To replace and legitimately close this issue i've updated #28836

**Comment by fabpot** [comment]
closing in favor of #28831

**Pull request closed without merging**